pkg/applicationserver/metadata/location_registry.go

```diff
@@ -0,0 +1,248 @@
+	if err != nil {
+		return nil, err
+	}
+	cl := ttnpb.NewEndDeviceRegistryClient(cc)
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+	dev, err := cl.Get(ctx, &ttnpb.GetEndDeviceRequest{
+		EndDeviceIdentifiers: ids,
+		FieldMask:            endDeviceLocationFieldMask,
+	}, c.WithClusterAuth())
+	if err != nil {
+		return nil, err
+	}
+	return dev.Locations, nil
+}
+
+// Merge implements EndDeviceLocationRegistry.
+func (c clusterEndDeviceLocationRegistry) Merge(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, update map[string]*ttnpb.Location) (map[string]*ttnpb.Location, error) {
+	cc, err := c.GetPeerConn(ctx, ttnpb.ClusterRole_ENTITY_REGISTRY, nil)
+	if err != nil {
+		return nil, err
+	}
+	cl := ttnpb.NewEndDeviceRegistryClient(cc)
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+	dev, err := cl.Get(ctx, &ttnpb.GetEndDeviceRequest{
+		EndDeviceIdentifiers: ids,
+		FieldMask:            endDeviceLocationFieldMask,
+	}, c.WithClusterAuth())
+	if err != nil {
+		return nil, err
+	}
+	if len(update) == 0 {
+		return dev.Locations, nil
+	}
+	if dev.Locations == nil {
+		dev.Locations = make(map[string]*ttnpb.Location, len(update))
+	}
+	for k, l := range update {
+		dev.Locations[k] = l
+	}
+	_, err = cl.Update(ctx, &ttnpb.UpdateEndDeviceRequest{
+		EndDevice: ttnpb.EndDevice{
+			EndDeviceIdentifiers: ids,
+			Locations:            dev.Locations,
+		},
+		FieldMask: endDeviceLocationFieldMask,
+	}, c.WithClusterAuth())
+	if err != nil {
+		return nil, err
+	}
+	return dev.Locations, nil
+}
+
+// NewClusterEndDeviceLocationRegistry returns an EndDeviceLocationRegistry connected to the Entity Registry.
+func NewClusterEndDeviceLocationRegistry(cluster ClusterPeerAccess, timeout time.Duration) EndDeviceLocationRegistry {
+	return &clusterEndDeviceLocationRegistry{
+		ClusterPeerAccess: cluster,
+		timeout:           timeout,
+	}
+}
+
+type cachedEndDeviceLocationRegistry struct {
+	registry EndDeviceLocationRegistry
+	cache    EndDeviceLocationCache
+
+	softTTL  time.Duration
+	hardTTL  time.Duration
+	errorTTL time.Duration
+
+	replicationPool workerpool.WorkerPool
+}
+
+// Get implements EndDeviceLocationRegistry.
+func (c *cachedEndDeviceLocationRegistry) Get(ctx context.Context, ids ttnpb.EndDeviceIdentifiers) (map[string]*ttnpb.Location, error) {
+	locations, ttl, err := c.cache.Get(ctx, ids)
+	switch {
+	case err != nil && !errors.IsNotFound(err):
+		return nil, err
+	case err != nil && errors.IsNotFound(err):
+		locations = nil
+	case err == nil:
+		window := c.hardTTL - c.softTTL
+		remaining := window - ttl
+		if remaining <= 0 {
+			return locations, nil
+		}
+		threshold := time.Duration(random.Int63n(int64(window)))
+		// remaining is the remaining window of the soft TTL in the (0, window) interval.
+		// threshold is a uniformly distributed duration in the [0, window) interval.
+		if remaining >= threshold {
+			return locations, nil
+		}
```

Comment:
Here is the soft TTL flow, that occurs on cache hit. `window` is the window of time in which a soft TTL asynchronous cache refresh can occur. `remaining` is how much of the window is left. At remaining `ttl`=`window`, the probability is 0% that we do an asynchronous cache retrieval, and at remaining `ttl`=`0` the probability is 100% that we do an asynchronous cache retrieval.

We don't block when this occurs - if the cache was hit (so we still have these values) we return them immediately. The cache refresh occurs in the background.